morango/utils/uuids.py

```diff
@@ -3,88 +3,146 @@
-    
+    id = UUIDField(max_length=32, primary_key=True)
+
+    class Meta:
+        abstract = True
+
     def calculate_uuid(self):
         """Should return a 32-digit hex UUID that is calculated as a function of the jointly
         unique fields on the model."""
-        
+
         # raise an error if no inputs to the UUID calculation were specified
         if self.uuid_input_fields is None:
-            raise NotImplemented("""You must define either a 'uuid_input_fields' attribute
+            raise NotImplementedError("""You must define either a 'uuid_input_fields' attribute
                 (with a tuple of field names) or override the 'calculate_uuid' method, on models
                 that inherit from UUIDModelMixin. If you want a fully random UUID, you can set
                 'uuid_input_fields' to the string 'RANDOM'.""")
-        
+
         # if the UUID has been set to be random, return a random UUID
         if self.uuid_input_fields == "RANDOM":
             return uuid.uuid4()
-            
+
         # if we got this far, uuid_input_fields should be a tuple
         assert isinstance(self.uuid_input_fields, tuple), "'uuid_input_fields' must either be a tuple or the string 'RANDOM'"
-        
+
         # calculate the input to the UUID function
         hashable_input_vals = []
         for field in self.uuid_input_fields:
             new_value = getattr(self, field)
             if new_value:
                 hashable_input_vals.append(str(new_value))
         hashable_input = ":".join(hashable_input_vals)
-        
+
         # if all the values were falsey, just return a random UUID, to avoid collisions
         if not hashable_input:
             return uuid.uuid4()
-        
+
         # compute the UUID as a function of the input values
         return uuid.uuid5(NAMESPACE_MORANGO, hashable_input)
 
     def save(self, *args, **kwargs):
-        
+
         if not self.id:
             self.id = self.calculate_uuid()
-            
+
         super(UUIDModelMixin, self).save(*args, **kwargs)
 
 
-class InstanceIDModel(models.Model, UUIDModelMixin):
-    
+class DatabaseIDModel(UUIDModelMixin):
+    """
+    Model to be used for tracking database ids.
+    """
+
+    uuid_input_fields = "RANDOM"
+
+    current = models.BooleanField(default=True)
+    date_generated = models.DateTimeField(default=timezone.now)
+    initial_instance_id = models.CharField(max_length=32, blank=True)
+
+
+class InstanceIDModel(UUIDModelMixin):
+
     uuid_input_fields = ("platform", "hostname", "sysversion", "macaddress")
-    
+
     platform = models.TextField()
     hostname = models.TextField()
     sysversion = models.TextField()
-    macaddress = models.CharField(blank=True)
-    
+    macaddress = models.CharField(max_length=20, blank=True)
+    database = models.ForeignKey(DatabaseIDModel)
+    counter = models.IntegerField(default=0)
+    current = models.BooleanField(default=True)
```

Comment:
We should also add a `db_path` field, which gets set to the path (on disk) of the current database -- that way, if we have two different databases on the same system, they won't get the same instance ID.

The value for this can be pulled from `settings.DATABASES["default"]["name"]` (which will work even if it's postgres or something else, as it will distinguish between multiple DBs by the DB/connection name). If it's sqlite, and the name doesn't start with a `/`, we should convert it to an absolute path.